docs/SETUP_PROVISIONING.md

```diff
@@ -0,0 +1,1505 @@
+The remainders of sections explain in more details how to install the tools mentioned here and give additional info on the process.
+
+### Pre-requisites: Setup pipeline and Install the tools
+
+1. You have implemented the instructions in `docs/SETUP_CI_CD_PIPELINE.md`
+2. Install AWS CLI by following AWS's instructions[here](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html).
+3. Install Terraform with `brew install terraform`
+4. Install Ansible with `brew install ansible`
+
+More details about the tools can be found in the "Tools" section futher down.
+
+### Setting up your Staging environment
+
+#### 1.Provision AWS infrastructure
+
+```
+$ rm -rf ops/infrastructure/envs/staging
+$ mkdir ops/infrastructure/envs/staging
+$ cd ops/infrastructure/envs/staging/
+$ ../../../scripts/tf_init.sh --project \<your gitlab project path\> --env staging --region \<your region\> --ssh-key \<path to your key\> --web-ec2-type t3.small --bastion-ec2-type t3.small
+$ AWS_PROFILE=gigadb terraform plan
+$ AWS_PROFILE=gigadb terraform apply
+$ AWS_PROFILE=gigadb terraform refresh
+```
+
+>**Note:** Core team members who are deploying on Upstream should use `AWS_PROFILE=upstream` or `AWS_PROFILE=upstreamAlt`
+
+#### 2. Connect provisioning with Gitlab pipeline
+
+```
+$ ../../../scripts/ansible_init.sh --env staging
+```
+This will save the information Terraform retrieved from the just created AWS resources into Gitlab variables.
+You can now start a Gitlab pipeline for the git branch or tag you want to deploy (the main branch is called `develop`).
+
+
+#### 3. Configure the infastructure
+ 
+```
+$ env TF_KEY_NAME=private_ip OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook  -i ../../inventories webapp_playbook.yml -e="gigadb_env=staging"
+$ env TF_KEY_NAME=private_ip OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook  -i ../../inventories files_playbook.yml -e="gigadb_env=staging"
+```
+
+#### 4. Deploy the web-based application and the FTP server
+
+Use the Gitlab Pipelines dashboard to build the application (by triggering manual job ``build_staging``).
+Once the production containers are build (in the Gitlab Containers Registry they are listed as production_\<service\>:\<environment\>),
+We can proceed with deployment to the target environment by triggering manual job ``sd_gigadb`` (staging).
+
+The application should be available at the url defined in $REMOTE_HOME_URL for the staging environment
+
+For the FTP, trigger on the Gitlab pipelione the build job `PureFtpdBuildStaging`, and upon success trigger the deployment with `PureFtpdDeployStaging`.
+
+>**Note:** The `build_staging` and `sd_gigadb` jobs are automated, so if you take too long to perform step 3, the jobs may run and fails.
+>That's fine, you can just trigger the jobs manually whenever you have completed step 3.
+
+#### 5. Configure the bastion server and deploy the back-office subsystems
```

Comment:
I needed to have bastion server already provisioned by its Ansible playbook otherwise I get this error when running `sd_gigadb`:
```
[21-Jul-2024 20:46:55 UTC] PHP Warning:  Unknown: php_network_getaddresses: getaddrinfo failed: Name or service not known in Unknown on line 0. Running Yii command generatefiletypes. CDbException: The table "file_type" for active record class "FileType" cannot be found in the database. in /var/www/vendor/yiisoft/yii/framework/db/ar/CActiveRecord.php:2391
```

Might be better to incorporate this part into section 3.

Reply:
Hi @pli888,
 
the problem  here is that are there parts of the bastion playbook that requires the `build_staging` and `sd_gigadb` to have run beforhand.

from the error you get it sounds like there is a  network  setup done in bastion playbook that should have been done in bootstrap playbook (run by `ansbile_init.sh`.

I'll create a ticket for that. In the meantime I'll make the  incantation to bastion playbook to appear in both section.

Reply:
Created a ticket #1985 for the above issue.

Reply:
Hi @pli888,

Instead of running the `bastion_playbook.yml` twice, we can use the `--tags` to control what tasks need to run when:

Before running the gitlab pipeline: 

env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml -e "gigadb_env=staging" --tags="fix-centos-eol-issues,setup-fail2ban,setup-docker-ce,restore-db-on-rds, create-metadata-bucket-credentials-file,setup-uploading-db-dump-to-s3,rclone-tool,node-exporter"


After running the gitlab pipeline:

env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml -e "gigadb_env=staging" --tags="files-url-updater,readme-tool,excel-spreadsheet-uploader,files-metadata-tool,load-latest-db"


I'll test those work and update the doc accordingly. 
It also offers us a strong hint on how to implement #1985